docs/strategy_reviewable.md

```diff
@@ -0,0 +1,255 @@
+4. **measure total** - produce a single overarching measurement for the entire program execution
+    - for a single instruction we can then have a **measure pro-rated** value, whereby we "split" the measurement total into contributions proportional to the count of each OPCODE instructions (**NOTE** this way of counting might be controversial and requires a special approach to the sample program set)
+    - for a single instruction we can then have a **measure inferred** value, whereby we deduct the _a priori_ known measurements of other instructions from the total. E.g. we have a value for `PUSH1` and for a `PUSH1, PUSH1, ADD` program we can infer a measurement of `ADD`.
+
+When we speak of measurements, we have the following values in mind (to be decided):
+1. **time** - wall clock time (the most universal and likely one)
+2. **cycles** - CPU cycles using TSC (`RDTSC`, `gotsc` etc.)
+
+### Individual measurement vs entire program execution measurement
+
+The main question regarding the measurement of (CPU-bound) computational cost of instructions is, whether it is the individual instruction execution to be measured or an entire program execution, consisting of multiple instructions.
+
+**Pros - individual instructions**
+
+[See here](https://htmlpreview.github.io/?https://github.com/imapp-pl/gas-cost-estimator/blob/master/src/analysis/exploration.nb.html) for preliminary results.
+
+Here we measure the computational cost of each iteration of the inner loop of the EVM interpreter.
+- measurement is more granular and allows us to analyze the measurements without worrying about statistical errors
+- in particular, it allows for granular discrimination between cost of OPCODEs located in various circumstances ("surroundings")
+
+**Cons**
+
+- need to be careful about timer precision and overhead (preliminary results try to account for it)
+
+**Pros - entire program measurement**
+
+[See for example here](See https://notes.ethereum.org/@chfast/benchmarking-evm-instructions)
+
+- impact of timer precision and overhead on the measurements is much less problematic
+- program execution more resembling the production environment, no tracing/debug calls
+
+**Cons**
+
+- couples the intrinsic effect of instruction circumstances with their statistical impact on the estimator outcome
+- must be more careful about program generation, to not have a dataset with adverse statistical properties (e.g. we'd need a very homogenous and random program generator, which is hard given the need to balance the stack)
+
+**Proposition**
+
+Use individual instructions as the estimator and the entire program measurement as validation.
+Our goal metric can be how closely does the estimate coming from individual measurements can model the cost of an entire program
+
+### Timer overhead adjustment
+
+It is natural, that the reading of a timer has itself an overhead. In a sequence:
+
+```
+t0 = timer.now()
```

Comment:
I would be much more comfortable if the measurement is placed in relation to the interpreter loop.

```
while (true):
  samples.push_back(timer.now())
  measured_code...
```

This make sure that the loop iteration is also accounted. But there is also additional overhead of storing the samples.